plugins/aver-hid/fu-aver-safeisp-device.c

```diff
@@ -0,0 +1,448 @@
+					      error)) {
+			g_prefix_error(error, "failed to send packet: ");
+			return FALSE;
+		}
+	}
+	if (res != NULL) {
+		if (!fu_hid_device_get_report(FU_HID_DEVICE(self),
+					      res->data[0],
+					      res->data,
+					      res->len,
+					      FU_AVER_SAFEISP_DEVICE_TIMEOUT,
+					      FU_HID_DEVICE_FLAG_USE_INTERRUPT_TRANSFER,
+					      error)) {
+			g_prefix_error(error, "failed to receive packet: ");
+			return FALSE;
+		}
+		g_debug("custom-isp-cmd: %s [0x%x]",
+			fu_aver_safeisp_custom_cmd_to_string(
+			    fu_struct_aver_safeisp_res_get_custom_cmd(res)),
+			fu_struct_aver_safeisp_res_get_custom_cmd(res));
+	}
+	return TRUE;
+}
+
+static gboolean
+fu_aver_safeisp_device_poll(FuDevice *device, GError **error)
+{
+	FuAverSafeispDevice *self = FU_AVER_SAFEISP_DEVICE(device);
+	g_autoptr(FuDeviceLocker) locker = NULL;
+	g_autoptr(GByteArray) req = fu_struct_aver_safeisp_req_new();
+	g_autoptr(GByteArray) res = fu_struct_aver_safeisp_res_new();
+
+	locker = fu_device_locker_new(device, error);
+	if (locker == NULL)
+		return FALSE;
+
+	fu_struct_aver_safeisp_req_set_custom_cmd(req, FU_AVER_SAFEISP_CUSTOM_CMD_GET_VERSION);
+	if (!fu_aver_safeisp_device_transfer(self, req, res, error))
+		return FALSE;
+
+	return TRUE;
+}
+
+static gboolean
+fu_aver_safeisp_device_ensure_version(FuAverSafeispDevice *self, GError **error)
+{
+	g_autofree gchar *ver = NULL;
+	g_autoptr(GByteArray) req = fu_struct_aver_safeisp_req_new();
+	g_autoptr(GByteArray) res = fu_struct_aver_safeisp_res_new();
+	fu_struct_aver_safeisp_req_set_custom_cmd(req, FU_AVER_SAFEISP_CUSTOM_CMD_GET_VERSION);
+	if (!fu_aver_safeisp_device_transfer(self, req, res, error))
+		return FALSE;
+	if (!fu_struct_aver_safeisp_res_device_version_validate(res->data, res->len, 0x0, error))
+		return FALSE;
+	ver =
+	    fu_strsafe((const gchar *)fu_struct_aver_safeisp_res_device_version_get_ver(res, NULL),
+		       FU_STRUCT_AVER_SAFEISP_RES_DEVICE_VERSION_SIZE_VER);
+	fu_device_set_version(FU_DEVICE(self), ver);
+	return TRUE;
+}
+
+static gboolean
+fu_aver_safeisp_device_setup(FuDevice *device, GError **error)
+{
+	FuAverSafeispDevice *self = FU_AVER_SAFEISP_DEVICE(device);
+
+	/* HidDevice->setup */
+	if (!FU_DEVICE_CLASS(fu_aver_safeisp_device_parent_class)->setup(device, error))
+		return FALSE;
+
+	/* using isp status requests as polling device requests */
+	if (!fu_aver_safeisp_device_poll(device, error))
+		return FALSE;
+
+	/* get the version from the hardware while open */
+	if (!fu_aver_safeisp_device_ensure_version(self, error))
+		return FALSE;
+
+	/* success */
+	return TRUE;
+}
+
+static FuFirmware *
+fu_aver_safeisp_device_prepare_firmware(FuDevice *device,
+					GInputStream *stream,
+					FwupdInstallFlags flags,
+					GError **error)
+{
+	g_autoptr(FuFirmware) firmware = fu_aver_hid_firmware_new();
+	if (!fu_firmware_parse_stream(firmware, stream, 0x0, flags, error))
+		return NULL;
+	return g_steal_pointer(&firmware);
+}
```

Comment:
I think you should be able to remove this entirely and use `fu_device_set_firmware_gtype(device, FU_TYPE_AVER_HID_FIRMWARE);` in `fu_aver_safeisp_device_init()` instead.